contracts/adapters/TributeNFT.sol

```diff
@@ -0,0 +1,267 @@
+    ) external override {
+        IVoting votingContract = IVoting(dao.getAdapterAddress(VOTING));
+        address sponsoredBy =
+            votingContract.getSenderAddress(
+                dao,
+                address(this),
+                data,
+                msg.sender
+            );
+        dao.sponsorProposal(proposalId, sponsoredBy);
+        votingContract.startNewVotingForProposal(dao, proposalId, data);
+    }
+
+    /**
+     * @notice Cancels a tribute proposal which marks it as processed and returns the NFT to the original ow.
+     * @dev Proposal id must exist.
+     * @dev Only proposals that have not already been sponsored can be cancelled.
+     * @dev Only proposer can cancel a tribute proposal.
+     * @param dao The DAO address.
+     * @param proposalId The proposal id.
+     */
+    function cancelProposal(DaoRegistry dao, bytes32 proposalId)
+        external
+        override
+    {
+        ProposalDetails storage proposal = proposals[address(dao)][proposalId];
+        require(proposal.id == proposalId, "proposal does not exist");
+        require(
+            !dao.getProposalFlag(
+                proposalId,
+                DaoRegistry.ProposalFlag.SPONSORED
+            ),
+            "proposal already sponsored"
+        );
+        require(
+            proposal.applicant == msg.sender,
+            "only the applicant can cancel a proposal"
+        );
+
+        dao.processProposal(proposalId);
+
+        NFTExtension nftExt = NFTExtension(dao.getExtensionAddress(NFT));
+        nftExt.returnNFT(
+            proposal.applicant,
+            proposal.nftAddr,
+            proposal.nftTokenId
+        );
+    }
+
+    /**
+     * @notice Processes the proposal to handle minting and exchange of DAO internal tokens for tribute tokens (passed vote) or the return the NFT to the original owner (failed vote).
+     * @dev Proposal id must exist.
+     * @dev Only proposals that have not already been processed are accepted.
+     * @dev Only sponsored proposals with completed voting are accepted.
+     * @dev ERC-721 tribute tokens must be registered with the DAO Bank.
+     * @param dao The DAO address.
+     * @param proposalId The proposal id.
+     */
+    function processProposal(DaoRegistry dao, bytes32 proposalId)
+        external
+        override
+    {
+        ProposalDetails storage proposal = proposals[address(dao)][proposalId];
+        require(proposal.id == proposalId, "proposal does not exist");
+
+        dao.processProposal(proposalId);
+
+        IVoting votingContract = IVoting(dao.getAdapterAddress(VOTING));
+        IVoting.VotingState voteResult =
+            votingContract.voteResult(dao, proposalId);
+
+        if (voteResult == IVoting.VotingState.PASS) {
+            _mintSharesToNewMember(
+                dao,
+                proposal.applicant,
+                proposal.nftAddr,
+                proposal.nftTokenId,
+                proposal.requestedShares
+            );
```

Comment:
so here the NFT should be moved to the NFTExtension and the owner should be the guild address (and not the person who sent it)

Reply:
sounds good. I will update that too.